reference/performance: add doc for join reorder (#2264)
pingcap/docs-cn: reference/performance/join-reorder.md

> ---

Reviewer: Please add this doc link into `TOC.md`.

> title: TiDB 当前的 Join Reorder 算法简介

Reviewer: ```suggestion
title: Join Reorder 算法简介
```

category: reference
---

> # Join Reorder

Reviewer: ```suggestion
# Join Reorder 算法简介
```
一级标题需要与上面元信息里的 title 保持一致。

> 在实际的的业务场景中，多个表的 Join 语句是非常常见的。而 Join 的执行效果会和各个表参与 Join 的顺序有关系。如 `select * from t, t1, t2 where t.a=t1.a and t2.a=t1.a`，这个 SQL 中可能的执行顺序有`t 和 t1 先做 Join，然后再和 t2 做 Join`以及`t1 和 t2 先做 Join，然后再和 t 做 Join` 两种情况。根据 `t` 和 `t2` 的数据量及数据分布，这两种执行顺序会有不同的性能表现。

Reviewer: ```suggestion
在实际的业务场景中，多个表的 Join 语句是很常见的，而 Join 的执行效果和各个表参与 Join 的顺序有关系。如 `select * from t, t1, t2 where t.a=t1.a and t2.a=t1.a`，这个 SQL 中可能的执行顺序有“t 和 t1 先做 Join，然后再和 t2 做 Join”以及“t1 和 t2 先做 Join，然后再和 t 做 Join”两种情况。根据 `t` 和 `t2` 的数据量及数据分布，这两种执行顺序会有不同的性能表现。
```

> 在实际的的业务场景中，多个表的 Join 语句是非常常见的。而 Join 的执行效果会和各个表参与 Join 的顺序有关系。如 `select * from t, t1, t2 where t.a=t1.a and t2.a=t1.a`，这个 SQL 中可能的执行顺序有`t 和 t1 先做 Join，然后再和 t2 做 Join`以及`t1 和 t2 先做 Join，然后再和 t 做 Join` 两种情况。根据 `t` 和 `t2` 的数据量及数据分布，这两种执行顺序会有不同的性能表现。

Reviewer 2: > 而 Join 的执行效果和各个表参与 Join 的顺序

“效果”用“效率”或者“性能”更好？（“效果”有一种结果会不一样的感觉）

> 因此需要优化器实现一种决定 Join 顺序的算法。目前 TiDB 中的算法是贪心算法。

Reviewer: ```suggestion
因此优化器需要实现一种决定 Join 顺序的算法。目前 TiDB 中使用的算法是 Join Reorder 算法，又称贪心算法。
```
Join Reorder 算法 和 贪心算法 是一个意思吗？文章的内容需要统一一下这两个词的用法。

> # 贪心算法简介

Reviewer: ```suggestion
## Join Reorder 算法实例
```
一篇文档只能有一个一级标题，即全文的标题。

> 以三个表 `t1, t2, t3` 的 Join 为例。首先我们会从拿到所有参与 Join 的节点，将他们按照行数多少从少到多排序

Reviewer: ```suggestion
以三个表 t1、t2、t3 的 Join 为例。首先获取所有参与 Join 的节点，将所有节点按照行数多少，从少到多进行排序。
```

> 以三个表 `t1, t2, t3` 的 Join 为例。首先我们会从拿到所有参与 Join 的节点，将他们按照行数多少从少到多排序

Reviewer 2: 这里用了 `t1, t2, t3` ，要不要把上面的例子也改成 `t1, t2, t3`?

> ![pic1](/media/join-reorder-1.png)

Reviewer: ```suggestion
![join-reorder-1](/media/join-reorder-1.png)
```
For SEO reasons, it's better to make the image link description self-explanatory. Please update all image link descriptions in this document.

Reviewer: 文档中还有几个其他的 link description 需要更新。 @winoros

> 之后我们选定其中最小的表。和其他两个表一次做 Join，看输出的结果集大小，选择更小的一侧作为选择。

Reviewer: ```suggestion
之后选定其中最小的表，将其与其他两个表分别做一次 Join，观察输出的结果集大小，选择其中结果更小的一对。
```

> ![pic2](/media/join-reorder-2.png)

Reviewer: ```suggestion
![join-reorder-2](/media/join-reorder-2.png)
```

> 然后进入下一轮的选择，如果这时是四个表那么就是继续比较输出结果集的大小进行选择。这里只有三个表，因此就直接得到了最终的 Join 结果。

Reviewer: ```suggestion
然后进入下一轮的选择，如果这时是四个表，那么就继续比较输出结果集的大小，进行选择。这里只有三个表，因此就直接得到了最终的 Join 结果。
```

![pic3](/media/join-reorder-3.png)

> 这个就是当前 TiDB 中使用的 Join reorder 算法。

Reviewer: ```suggestion
以上就是当前 TiDB 中使用的 Join reorder 算法。
```

> ![pic3](/media/join-reorder-3.png)

Reviewer: ```suggestion
![join-reorder-3](/media/join-reorder-3.png)
```

> # 限制

Reviewer: ```suggestion
## Join reorder 算法限制
```
